src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -1189,33 +1192,33 @@ private void ConsumeFromBuffer(int count)
             _receiveBufferOffset += count;
         }
 
-        private async Task EnsureBufferContainsAsync(int minimumRequiredBytes, CancellationToken cancellationToken, bool throwOnPrematureClosure = true)
+        private ValueTask EnsureBufferContainsAsync(int minimumRequiredBytes, CancellationToken cancellationToken, bool throwOnPrematureClosure = true)
         {
             Debug.Assert(minimumRequiredBytes <= _receiveBuffer.Length, $"Requested number of bytes {minimumRequiredBytes} must not exceed {_receiveBuffer.Length}");
-
-            // If we don't have enough data in the buffer to satisfy the minimum required, read some more.
-            if (_receiveBufferCount < minimumRequiredBytes)
+            try
             {
-                // If there's any data in the buffer, shift it down.  
-                if (_receiveBufferCount > 0)
-                {
-                    _receiveBuffer.Span.Slice(_receiveBufferOffset, _receiveBufferCount).CopyTo(_receiveBuffer.Span);
-                }
-                _receiveBufferOffset = 0;
-
-                // While we don't have enough data, read more.
-                while (_receiveBufferCount < minimumRequiredBytes)
+                // If we don't have enough data in the buffer to satisfy the minimum required, read some more.
+                if (_receiveBufferCount < minimumRequiredBytes)
                 {
-                    int numRead = await _stream.ReadAsync(_receiveBuffer.Slice(_receiveBufferCount, _receiveBuffer.Length - _receiveBufferCount), cancellationToken).ConfigureAwait(false);
-                    Debug.Assert(numRead >= 0, $"Expected non-negative bytes read, got {numRead}");
-                    if (numRead <= 0)
+                    // If there's any data in the buffer, shift it down.  
+                    if (_receiveBufferCount > 0)
                     {
-                        ThrowIfEOFUnexpected(throwOnPrematureClosure);
-                        break;
+                        _receiveBuffer.Span.Slice(_receiveBufferOffset, _receiveBufferCount).CopyTo(_receiveBuffer.Span);
                     }
-                    _receiveBufferCount += numRead;
+                    _receiveBufferOffset = 0;
+
+                    EnsureBuffer ensureBuffer = (_ensureBuffer ??= new EnsureBuffer(this));
+                    return ensureBuffer.EnsureBufferContainsAsync(minimumRequiredBytes, cancellationToken, throwOnPrematureClosure);
+                }
+                else
+                {
+                    return default;
                 }
             }
+            catch (Exception ex)
+            {
+                return new ValueTask(Task.FromException(ex));
```

Comment:
Go with raw exceptions for ValueTask rather than wrapping in Task?

Reply:
In this case, yes. Not because it's ValueTask, but because a) there shouldn't be any exceptions, b) it's private and we control all the call sites, and c) those call sites won't be able to tell the difference.

Reply:
Makes sense :) Changed